lib/kv/manager/src/storage.rs

```diff
@@ -0,0 +1,293 @@
+//!
+//! This module provides traits and implementations for managing storage of KV blocks.
+//! It handles system memory, pinned memory, device memory, and remote (NIXL) storage,
+//! with a focus on safety and performance.
+
+#[cfg(feature = "cuda")]
+mod cuda;
+
+#[cfg(feature = "cuda")]
+pub use cuda::*;
+
+use std::{
+    alloc::{alloc_zeroed, dealloc, Layout},
+    fmt::Debug,
+    ptr::NonNull,
+    sync::Arc,
+};
+use thiserror::Error;
+
+/// Result type for storage operations
+pub type StorageResult<T> = std::result::Result<T, StorageError>;
+
+/// Represents the type of storage used for a block
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum StorageType {
+    #[cfg(feature = "cuda")]
+    /// CUDA device storage
+    Device(Arc<CudaContext>),
+
+    /// Pinned host storage
+    Pinned,
+
+    /// System memory storage
+    System,
+
+    /// Null storage (for testing)
+    Null,
+}
+
+/// Errors that can occur during storage operations
+#[derive(Debug, Error)]
+#[allow(missing_docs)]
+pub enum StorageError {
+    #[error("Storage allocation failed: {0}")]
+    AllocationFailed(String),
+
+    #[error("Storage not accessible: {0}")]
+    NotAccessible(String),
+
+    #[error("Invalid storage configuration: {0}")]
+    InvalidConfig(String),
+
+    #[error("Storage operation failed: {0}")]
+    OperationFailed(String),
+
+    #[cfg(feature = "cuda")]
+    #[error("CUDA error: {0}")]
+    Cuda(#[from] DriverError),
+}
+
+/// Result type for storage operations
+pub type Result<T> = std::result::Result<T, StorageError>;
+
+/// Core storage trait that provides access to memory regions
+pub trait Storage: Debug + Send + Sync + 'static {
+    /// Returns the type of storage
+    fn storage_type(&self) -> StorageType;
+
+    /// Returns the address of the storage
+    fn addr(&self) -> u64;
+
+    /// Returns the total size of the storage in bytes
+    fn size(&self) -> usize;
+
+    /// Returns true if the storage is accessible by the host/cpu portion
+    /// of the application.
+    fn is_host_accessible(&self) -> bool;
+
+    /// Get a raw pointer to the storage
+    ///
+    /// # Safety
+    /// The caller must ensure:
+    /// - The pointer is not used after the storage is dropped
+    /// - Access patterns respect the storage's thread safety model
+    unsafe fn as_ptr(&self) -> Option<*const u8>;
+
+    /// Get a raw mutable pointer to the storage
+    ///
+    /// # Safety
+    /// The caller must ensure:
+    /// - The pointer is not used after the storage is dropped
+    /// - No other references exist while the pointer is in use
+    /// - Access patterns respect the storage's thread safety model
+    unsafe fn as_mut_ptr(&mut self) -> Option<*mut u8>;
+}
+
+/// Trait for types that can allocate specific Storage implementations.
+pub trait StorageAllocator<S: Storage>: Send + Sync {
+    /// Allocate storage of the specific type `S` with the given size in bytes.
+    fn allocate(&self, size: usize) -> Result<S>;
+}
```

Comment:
Is this the same as https://doc.rust-lang.org/stable/core/alloc/trait.Allocator.html